File: src/dmp_param/TauSystem.cpp
#include "dmp/dmp_param/TauSystem.h"

namespace dmp {

TauSystem::TauSystem()
    : tau_base(0.0),
      tau_reference(0.0),
      tau_coupling(NULL),
      rt_assertor(NULL) {}

TauSystem::TauSystem(double tau_base_init, RealTimeAssertor* real_time_assertor,
                     double tau_ref, std::vector<TauCoupling*>* tau_couplers)
    : tau_base(tau_base_init),
      tau_reference(tau_ref),
      tau_coupling(tau_couplers),
      rt_assertor(real_time_assertor) {}

bool TauSystem::isValid() {
  if (rt_assert(tau_base >= MIN_TAU) == false) {
    return false;
  }
  if (rt_assert(tau_reference >= MIN_TAU) == false) {
    return false;
  }
  return true;
}

bool TauSystem::setTauBase(double tau_base_new) {
  // input checking
  if (rt_assert(tau_base_new >= MIN_TAU) ==
      false)  // new tau_base is too small/does NOT make sense
  {
    return false;
  }

  tau_base = tau_base_new;
  return true;
}

bool TauSystem::getTauBase(double& tau_base_return) {
  // pre-condition(s) checking
  if (rt_assert(TauSystem::isValid()) == false) {
    return false;
  }

  tau_base_return = tau_base;
  return true;
}

bool TauSystem::getTauRelative(double& tau_relative) {
  // pre-condition(s) checking
  if (rt_assert(TauSystem::isValid()) == false) {
    return false;
  }

  double C_tau = 0.0;
  if (rt_assert(TauSystem::getCouplingTerm(C_tau)) == false) {
    return false;
  }

  tau_relative = (1.0 + C_tau) * (tau_base / tau_reference);
  if (rt_assert((tau_relative * tau_reference) >= MIN_TAU) ==
      false)  // too small/does NOT make sense
  {
    tau_relative = 0.0;
    return false;
  }

  return true;
}

bool TauSystem::getCouplingTerm(double& accumulated_ctau) {
  accumulated_ctau = 0.0;

  if (tau_coupling) {
    for (int i = 0; i < tau_coupling->size(); ++i) {
      if ((*tau_coupling)[i] == NULL) {
        continue;
      }

      double ctau = 0.0;
      if (rt_assert((*tau_coupling)[i]->getValue(ctau)) == false) {
        return false;
      }
      if (rt_assert(std::isnan(ctau) == 0) == false) {
        return false;
      }

      accumulated_ctau += ctau;
    }
  }

  return true;
}

double TauSystem::getTauReference() { return tau_reference; }

TauSystem::~TauSystem() {}

}  // namespace dmp

File: python/dmp_param/TauSystem.py
#!/usr/bin/env python2
# -*- coding: utf-8 -*-
"""
Created on Mon Oct 30 19:00:00 2017

@author: gsutanto
"""

import numpy as np
import os
import sys
sys.path.append(os.path.join(os.path.dirname(__file__), '../dmp_coupling/base/'))
sys.path.append(os.path.join(os.path.dirname(__file__), '../utilities/'))
from definitions_base import *
from TauCoupling import *

class TauSystem:
    'Class defining computation of tau parameter for canonical system, transformation system, and goal evolution system, which sometime is coupled with an external variable.'
    
    def __init__(self, tau_base_init, tau_ref=0.5, tau_couplers_list=[], name=""):
        self.name = name
        self.tau_base = tau_base_init
        self.tau_reference = tau_ref
        self.tau_couplers_list = tau_couplers_list
    
    def isValid(self):
        assert (self.tau_base >= MIN_TAU), "TauSystem.tau_base=" + str(self.tau_base) + " < MIN_TAU"
        assert (self.tau_reference >= MIN_TAU), "TauSystem.tau_reference=" + str(self.tau_reference) + " < MIN_TAU"
        return True
    
    def setTauBase(self, tau_base_new):
        assert (tau_base_new >= MIN_TAU), "tau_base_new have to be >= MIN_TAU"
        self.tau_base = tau_base_new
        return None
    
    def getTauRelative(self):
        assert (self.isValid()), "TauSystem is invalid!"
        C_tau = self.getCouplingTerm()
        tau_relative = (1.0 + C_tau) * (self.tau_base * 1.0 / self.tau_reference)
        assert (tau_relative * self.tau_reference >= MIN_TAU), 'tau_relative is too small!'
        return tau_relative
    
    def getCouplingTerm(self):
        accumulated_ctau = 0.0
        for tau_coupler_idx in range(len(self.tau_couplers_list)):
            ctau = self.tau_couplers_list[tau_coupler_idx].getValue()
            assert (np.isnan(ctau) == False), 'ctau['+str(tau_coupler_idx)+'] is NaN!'
            accumulated_ctau = accumulated_ctau + ctau
        return accumulated_ctau